From: Florian Westphal <fw@strlen.de>
Date: Tue, 30 Jun 2015 22:27:51 +0200
Subject: netfilter: bridge: don't leak skb in error paths
Patch-mainline: v4.2-rc3
Git-commit: dd302b59bde0149c20df7278c0d36c765e66afbd
References: bsc#982544

br_nf_dev_queue_xmit must free skb in its error path.
NF_DROP is misleading -- its an okfn, not a netfilter hook.

Fixes: 462fb2af9788a ("bridge : Sanitize skb before it enters the IP stack")
Fixes: efb6de9b4ba00 ("netfilter: bridge: forward IPv6 fragmented packets")
Signed-off-by: Florian Westphal <fw@strlen.de>
Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>
Signed-off-by: Michal Kubecek <mkubecek@suse.cz>

---
 net/bridge/br_netfilter.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/net/bridge/br_netfilter.c b/net/bridge/br_netfilter.c
index 55718e32a2ae..4c8b681d5887 100644
--- a/net/bridge/br_netfilter.c
+++ b/net/bridge/br_netfilter.c
@@ -850,7 +850,7 @@ static int br_nf_dev_queue_xmit(struct sk_buff *skb)
 #if defined(CONFIG_NF_DEFRAG_IPV4) || defined(CONFIG_NF_DEFRAG_IPV4_MODULE)
 	if (skb->protocol == htons(ETH_P_IP)) {
 		if (br_parse_ip_options(skb))
-			return NF_DROP;
+			goto drop;
 		return ip_fragment(skb, br_dev_queue_push_xmit);
 	}
 #endif
@@ -859,15 +859,19 @@ static int br_nf_dev_queue_xmit(struct sk_buff *skb)
 		const struct nf_ipv6_ops *v6ops = nf_get_ipv6_ops();
 
 		if (br_validate_ipv6(skb))
-			return NF_DROP;
+			goto drop;
 		if (v6ops)
 			return v6ops->fragment(skb, br_dev_queue_push_xmit);
-		else
-			return -EMSGSIZE;
+
+		kfree_skb(skb);
+		return -EMSGSIZE;
 	}
 #endif
 
 	return br_dev_queue_push_xmit(skb);
+drop:
+	kfree_skb(skb);
+	return 0;
 }
 #else
 static int br_nf_dev_queue_xmit(struct sk_buff *skb)
-- 
2.9.0

